packages/api/src/incentives/surge/marginAccountDbService.ts

```diff
@@ -0,0 +1,256 @@
+          }).pipe(
+            Effect.map((result) => {
+              if (result.length === 0) {
+                return null;
+              }
+              const record = result[0]!;
+              // Return trading account if available, otherwise fall back to recovery account
+              return (
+                record.tradingAccountAddress || record.recoveryAccountAddress
+              );
+            }),
+          );
+        }),
+
+        // Get the total count of unique margin accounts
+        getCount: () =>
+          Effect.gen(function* () {
+            return yield* Effect.tryPromise({
+              try: () =>
+                db
+                  .selectDistinctOn([marginAccounts.marginAccountAddress])
+                  .from(marginAccounts),
+              catch: (error) => new DbError(error),
+            }).pipe(Effect.map((results) => results.length));
+          }),
+
+        // Get the collateral account address for a margin account at a specific state version
+        // Returns collateral account, or falls back to recovery account if collateral account is null
+        // Returns null if no ownership record exists at or before the target state version
+        getCollateralAccountAtStateVersion: Effect.fn(function* (
+          marginAccountAddress: string,
+          targetStateVersion: number,
+        ) {
+          return yield* Effect.tryPromise({
+            try: () =>
+              db
+                .select({
+                  collateralAccountAddress:
+                    marginAccounts.collateralAccountAddress,
+                  recoveryAccountAddress: marginAccounts.recoveryAccountAddress,
+                })
+                .from(marginAccounts)
+                .where(
+                  and(
+                    eq(
+                      marginAccounts.marginAccountAddress,
+                      marginAccountAddress,
+                    ),
+                    lte(marginAccounts.stateVersion, targetStateVersion),
+                  ),
+                )
+                .orderBy(desc(marginAccounts.stateVersion))
+                .limit(1),
+            catch: (error) => new DbError(error),
+          }).pipe(
+            Effect.map((result) => {
+              if (result.length === 0) {
+                return null;
+              }
+              const record = result[0]!;
+              // Return collateral account if available, otherwise fall back to recovery account
+              return (
+                record.collateralAccountAddress || record.recoveryAccountAddress
+              );
+            }),
+          );
+        }),
+
+        // Get margin accounts by collateral addresses for efficient batch processing
+        // Returns the latest margin account for each collateral address at the target state version
+        // If a margin account has no collateral account, falls back to recovery account
+        getMarginAccountsByCollateralAddresses: Effect.fn(function* (
```

Comment:
add some test here to ensure that it's getting the correct margin account at a given state version